### PR TITLE
Make manim installable using setuptools (setup.py)

### PR DESCRIPTION
**Pull request by JohnAZoidberg**
I want to package manim for my distribution and made changes here so that it's easier for me and others to package it.

Now the project supports the standard Python packaging method of setuptools and we can install it using just `python3 -m pip install .` (even without a distribution packaging it). After installing we can use it in any directory - not just the source repository.

I'm still wondering which are the files that people are supposed to run.
`manim.py` is obviously the main one but does it allow access to all the functionality?
`stage_scenes.py` looks like it does something useful, but I can't figure out exactly what. See: #515 
`manimlib/extract_scene.py` has an `if __name__ == "__main__"` but that's broken because it doesn't call `main()` with all its parameters.

If you like the idea of it being packaged in distribution's package managers I'd also appreciate if you made a release on GitHub with a version number.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env python
```
Comment (FRidh) [inline setup.py]: setup.py is typically not executed, so remove the shebang

Comment (JohnAZoidberg) [reply setup.py]: Yeah I guess you're right but they have it in their example: https://docs.python.org/3.7/distutils/setupscript.html
Kenneth Reitz has it too, in his tutorial on `setup.py`: https://github.com/kennethreitz/setup.py/blob/master/setup.py

Comment (FRidh) [reply setup.py]: ugh...this makes the assumption `python` exists, which isn't always the case (e.g. when only python3 is around an no symlink exists, which is how a certain PEP describes how it should be)

**Comment by eulertour** [comment]
This is a good idea for manim's development into a real library. A few things to note in order to avoid problems later:

1. Empty `__init__.py` files are redundant in python3.3+, so 5369843 should be reverted. The call to `find_packages` should be replaced with `find_namespace_packages`.
2. `big_ol_pile_of_manim_imports.py` should be added to `scripts`. Really, `big_ol_pile_of_manim_imports.py` should be moved into the library as `manimlib.imports`, but that would be a _very_ breaking change, so it will require at least notifying @3b1b.
3. The version should be 0.1 if not 0.0.1. We don't want to give the impression that this is a stable library yet, and when we do start cutting releases we'll probably end up going through version numbers very quickly.

Regarding your question, everything related to creating video should be accessible through `manim.py`. That said, I don't know what `stage_scenes.py` does.

**Comment by JohnAZoidberg** [comment]
>     1. Empty `__init__.py` files are redundant in python3.3+, so [5369843](https://github.com/3b1b/manim/commit/5369843667f32e0d3a882ae0d98894ecb956eca5) should be reverted. The call to `find_packages` should be replaced with `find_namespace_packages`.

Oh thanks, didn't know that!

>     2. `big_ol_pile_of_manim_imports.py` should be added to `scripts`. Really, `big_ol_pile_of_manim_imports.py` should be moved into the library as `manimlib.imports`, but that would be a _very_ breaking change, so it will require at least notifying @3b1b.

Yes, it should be a part of `manimlib` and yes that's quite breaking therefore I didn't attempt to do it.

I'm not sure about adding it to `scripts` - it's not really a script and it doesn't help to have that installed as a script (or does that make it importable? I'm not sure).
I thought of it more like an example file that you can copy to your project if you wanted to.
In the package I'm writing for my distro I add it do the `PYTHONPATH` would that also happen in `scripts`?

Oh and `manim.py` should probably also be in `manimlib` because that's the recommended structure for Python packages - to have everything in a package.

>     3. The version should be 0.1 if not 0.0.1. We don't want to give the impression that this is a stable library yet, and when we do start cutting releases we'll probably end up going through version numbers very quickly.

Yup. Seems like someone already stole the `manim` name on [Pypi](https://pypi.org/project/manim/) but I'm sure you could get that one because they're not really using it.

**Comment by eulertour** [comment]
> I'm not sure about adding it to scripts - it's not really a script and it doesn't help to have that installed as a script (or does that make it importable? I'm not sure).

I went through the install in a docker container and `big_ol_....` wasn't importable unless it was either moved into the library or added to scripts. I agree that it seems pretty hacky, but then again having one file separate from the rest of the library is pretty hacky in itself. There's a good chance I just don't know enough about setuptools to do it the "right" way.

> Oh and manim.py should probably also be in manimlib because that's the recommended structure for Python packages - to have everything in a package.

It seems that you have some prior experience with setuptools, or have at least done some research into  it. Can you link a source for this recommendation, or any other best practices you know of?

**Comment by JohnAZoidberg** [comment]
> I went through the install in a docker container and `big_ol_....` wasn't importable unless it was either moved into the library or added to scripts. I agree that it seems pretty hacky, but then again having one file separate from the rest of the library is pretty hacky in itself. There's a good chance I just don't know enough about setuptools to do it the "right" way.

Yes, tried it too, works :)

> > Oh and manim.py should probably also be in manimlib because that's the recommended structure for Python packages - to have everything in a package.
> 
> It seems that you have some prior experience with setuptools, or have at least done some research into it. Can you link a source for this recommendation, or any other best practices you know of?

Everyone still always recommends `__init__.py` but `manim` doesn't try to support old Python versions, right?
See the [documentation](https://setuptools.readthedocs.io/en/latest/setuptools.html#automatic-script-creation) of setuptools for why `entry_points` if preferable to `scripts`.
That already assumes that the script is contained in some module. You cannot specify just a file, only a method in a module.
Also see: https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html

**Pull request merged**